proposals/Imperative-Shadow-DOM-Distribution-API.md

```diff
@@ -1,48 +1,95 @@
+ 
+#### Case 1:  Slot attributes is required in markup.
+```html
+<shadow-host>
+  <div slot=slot1></div>
+  <div slot=slot2></div>
+</shadow-host>
+ ```
+Some people would see this as a kind of *ugly* markup.
+Shadow DOM v1 can't explain how `<summary>/<details>` elements can be implemented on the top of the current Web Components technology stack, given that `<details>` element doesn't need slot= attribute.
+Blink has a special logic for some built-in elements to control node-to-slot mapping.
+ 
+#### Case 2: Slot based on condition. 
+```html
+<custom-tab show-panel=”2”>
+   <tab-panel></tab-panel>
+   <tab-panel></tab-panel>
+   <tab-panel></tab-panel>
+</custom-tab>
+ ```
+The second issue is that component creators can't change the slotting behavior based on condition. In the above markup, the `<custom-tab>` component can't implement which `<tab-panel>` to show based on its attribute `show-panel`.
+
+## Imperative Slotting API
+
+The imperative slotting API allows the developer to explicitly set the [assigned nodes] for a slot element.
 
-The straw-man proposal for Imperative Shadow DOM distribution API.
-The context is: [https://github.com/whatwg/html/issues/3534](https://github.com/whatwg/html/issues/3534)
-
-# Declarative API and Imperative API should be exclusively used in each shadow tree
-
-Mixing declarative API and imperative API would be troublesome and can be the cause of confusions for web developers.
-We *can* invent complex rules, however, no one wants to remember complex rules. Also, supporting both in the same shadow tree would make a browser engine complex, which I don't want.
-
-Thus, we don't allow mixing declarative API and imperative API in the same shadow tree.
-Web developers have to show their *opt-in* to use imperative API for each shadow tree.
+[assigned nodes]: https://dom.spec.whatwg.org/#slot-assigned-nodes
 
-A shadow root has an associated *slotting*. Web developers can set shadow root's *slotting* to *manual* by specifying it in attachShadow:
+An `HTMLSlotElement` gets a new API, called `assign(sequence<Node> nodes)` (tentative name). Nodes are an ordered list of `slot's` host light DOM children. 
+ 
+#### Example Usage
+The following code demonstrate how to use the new API to solve [Case 2](#Case-2:-Slot-based-on-condition) of the Motivation section above.
 
 ```js
-const sr = attachShadow({ mode: ..., (optional) slotting: 'manual' })
+class CustomTab extends HTMLElement {
+    constructor() {
+        super();
+        const template = `
+            <div class="custom-tab">
+                <slot></slot>
+            </div>`;
+        const shadowRoot = this.attachShadow({mode: 'open', slotAssignment: 'manual'});
+        const div = document.createElement('div');
+        div.innerHTML = template;
+        shadowRoot.appendChild(div);
+    }
+    connectedCallback() {
```

Comment:
You might want to add a comment explaining that a production example would also use attributeChangeCallback to watch for changes to the `show-panel` attribute and re-run this code.

You could also just update the example to do that, but I'm not sure how much complexity it would add; perhaps it would be distracting.

Reply:
I've updated my example to include attributeChangedCallback.
Thanks.